Common/Orders/Slippage/MarketImpactSlippageModel.cs

```diff
@@ -0,0 +1,221 @@
+            // expected valid time for impact (+ half an hour)
+            var timePost = time + 1d / 13d;
+            // normalized volume of execution
+            var nu = (double)order.AbsoluteQuantity / time / symbolData.AvgVolume;
+            // liquidity adjustment for temporary market impact, if any
+            var liquidityAdjustment = asset.Fundamentals != null ?
+                                      Math.Pow(asset.Fundamentals.CompanyProfile.SharesOutstanding / symbolData.AvgVolume, _delta) :
+                                      1d;
+            // noise adjustment factor
+            var noise = symbolData.Sigma * Math.Sqrt(timePost);
+
+            // temporary market impact
+            var permImpact = symbolData.Sigma * time * G(nu) * liquidityAdjustment + SampleGaussian(_random) * noise;
+            // permanent market impact
+            var tempImpact = symbolData.Sigma * H(nu) + SampleGaussian(_random) * noise + permImpact * 0.5d;
+            // realized market impact
+            var realizedImpact = tempImpact + permImpact;
+
+            // estimate the slippage by temporary impact
+            return SlippageFromImpactEstimation(realizedImpact);
+        }
+
+        /// <summary>
+        /// The permanent market impact function
+        /// </summary>
+        /// <param name="absOrderQuantity">The absolute, normalized order quantity</param>
+        /// <return>Unadjusted permanent market impact factor</return>
+        protected double G(double absOrderQuantity)
+        {
+            return _gamma * Math.Pow((double)absOrderQuantity, _alpha);
+        }
+
+        /// <summary>
+        /// The temporary market impact function
+        /// </summary>
+        /// <param name="absOrderQuantity">The absolute, normalized order quantity</param>
+        /// <return>Unadjusted temporary market impact factor</return>
+        protected double H(double absOrderQuantity)
+        {
+            return _eta * Math.Pow((double)absOrderQuantity, _beta);
+        }
+
+        /// <summary>
+        /// The temporary market impact function
+        /// </summary>
+        /// <param name="tempImpact">The temporary market impact</param>
+        /// <return>Slippage estimation</return>
+        protected virtual decimal SlippageFromImpactEstimation(double tempImpact)
+        {
+            // We assume an exponential distribution on the order book
+            // the order with high volume of "no slippage/impact" zone as 0, mid point of max slippage as infinity
+            // we use lambda=2.5 as parameter, so the average slippage will be (1 - 1/2.5) of the impact
+            return Convert.ToDecimal(tempImpact) * 0.6m;
+        }
+
+        private double SampleGaussian(Random random, double mean = 0d, double stddev = 1d)
+        {
+            double x1 = 1 - random.NextDouble();
+            double x2 = 1 - random.NextDouble();
```

Comment:
This method has a bug. Its result cannot be negative, or the slippage is negative.
We need to ensure that x1 is greater than x2:
```csharp
if (x2 > x1) (x1, x2) = (x2, x1);
```
also, all `double` should be replaced by `var` to meet the repo style.